server: Add system.sessions table

mrtracy opened the pull request:
Add system.sessions table, which will be used to create user login
sessions for the Admin UI. Includes:

+ Definition of the new table
+ A backwards-compatible migration to add the table
+ Added a case to `TestSystemTableLiterals` to verify the new table
+ Updated a multitude of tests which explictly check the system schema,
they now account for the new table.

Commit:
server: Add system.web_sessions table

Add system.web_sessions table, which will be used to create user login
sessions for the Admin UI. Includes:

+ Definition of the new table
+ A backwards-compatible migration to add the table
+ Added a case to `TestSystemTableLiterals` to verify the new table
+ Updated a multitude of tests which explictly check the system schema,
they now account for the new table.

A possible oddity in this is that the new table has a TableID of 19,
while the previous real table had ID 15. This is because IDs 16, 17 and
18 were reserved to store zone configuration for non-table ranges.
This oddity causes 4 additional initial ranges to be created on the
server (due to the behavior of `SystemConfig.ComputeSplitKey()`
instead of just 1.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/17171)
<!-- Reviewable:end -->

benesch commented on pkg/sql/sqlbase/system.go:
```diff
@@ -123,6 +123,21 @@ CREATE TABLE system.jobs (
+	"auditInfo"    STRING,
+	INDEX("expiresAt"),
+	INDEX("createdAt"),
+	FAMILY(id, "hashedSecret", username, "createdAt", "expiresAt", "revokedAt", "lastUsedAt", "auditInfo")
```
Now that we're using Postgres case-folding rules, seems like we should switch to Postgres-style snake_case identifiers so we don't have to quote every multi-word column name. 

There's a consistency argument here too, though. I don't feel particularly strongly.

mrtracy replied:
Let's just as @bdarnell and he can make the call, I also see the merits and disadvantages of both approaches.  I went for the consistency angle.

vivekmenezes replied:
Can you document some of the fields as well as discuss why you need the indexes

benesch commented on pkg/keys/constants.go:
```diff
@@ -295,6 +295,7 @@ const (
 	RangeEventTableID = 13
 	UITableID         = 14
 	JobsTableID       = 15
+	SessionsTableID   = 19
```
19?

mrtracy replied:
Yeah, 16/17/18 are reserved to reference configuration for non-table entities.

I've updated this section to make it more apparent why this is 19.

benesch commented on pkg/storage/client_replica_test.go:
```diff
@@ -1518,26 +1518,11 @@ func TestSystemZoneConfigs(t *testing.T) {
-			t.Fatalf("failed to split at key %s: %s", key, err)
-		}
-	}
-
```
Are these always presplit these days?

mrtracy replied:
They are always presplit as a direct result of this checkin. System table IDs 16, 17 and 18 were reserved to hold custom zone config information for non-table ranges (Meta, time series, system), so the new table has to get ID 19. The logic in `ComputeSplitKey` assumes that table IDs are sequential, so now that it has table 19 it creates splits for 16, 17 and 18, even though they aren't actually table.

The real question I have, now that I think about it more, is why were these ranges being split before?  This test says "These ranges may get split at an inopportune time", but these are not real tables, so I don't think they ever should have been split at any point. @a-robinson Can you shed some light on that?

In the meantime, I've updated the commit message to provide more detail around this oddity.

a-robinson replied:
Looking at the code now, I'm not sure why they would have been split before. I remember having to add this bit because occasionally splits were needed when stressing the test, but don't remember why they were needed.

benesch reviewed (commented):
I realize I missed the boat in #16829, but `system.sessions` really makes me think of SQL sessions, as in `SHOW SESSIONS`. I wonder if this table would be better named `system.grpc_sessions`.

mrtracy commented:
As for the name of the table, I would be fine with something like "web_sessions", "auth_sessions", "token_sessions" or "cookie_sessions", but "grpc_sessions" is too specific to a technology.

Also, is "SHOW SESSIONS" a CockroachDB specific command? It doesn't appear to be the language used by any other database system.

benesch commented:
`SHOW SESSIONS`is Cockroach-specific, yup. Though I think MySQL and Postgres, at least, have the concept of "session" variables that last only while your connection is alive.

vivekmenezes commented on pkg/keys/constants.go:
```diff
@@ -287,19 +287,17 @@ const (
+	JobsTableID        = 15
 	MetaRangesID       = 16
 	SystemRangesID     = 17
 	TimeseriesRangesID = 18
```
these range related ids are not exported right?

vivekmenezes reviewed (commented):
Some drive by comments

benesch commented:
Reviewed 10 of 11 files at r1, 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[pkg/keys/constants.go, line 301 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17171#-Kpoy0xBf4S6y4GgCQAh:-Kpoy0xBf4S6y4GgCQAi:bxsmoqx) ([raw file](https://github.com/cockroachdb/cockroach/blob/c5e1cb0eb9eb4bb61ab774fb25c7369784c04ee0/pkg/keys/constants.go#L301)):*
> ```Go
> 	MetaRangesID       = 16
> 	SystemRangesID     = 17
> 	TimeseriesRangesID = 18
> ```

Thanks, @mrtracy, this is much clearer.

---

*[pkg/keys/constants.go, line 301 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17171#-Kpd8KqkP6xcBEgs4p2d-r2-301:-Kpoy7ktVX-P_FHIiS0C:b-w9i92t) ([raw file](https://github.com/cockroachdb/cockroach/blob/c5e1cb0eb9eb4bb61ab774fb25c7369784c04ee0/pkg/keys/constants.go#L301)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

these range related ids are not exported right?
</blockquote></details>

`MetaRangesID`, at least, needs to be exported for `./pkg/config` and `./pkg/storage`, if that's what you mean?

---

*[pkg/sql/sqlbase/system.go, line 139 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17171#-Kpd8KqkP6xcBEgs4p2k-r1-139:-KpoyO9n8QM_BNw5KMTw:b-1swmbg) ([raw file](https://github.com/cockroachdb/cockroach/blob/81c3b2e4d025f7a44b4d305d54b4ef348ddd8dd2/pkg/sql/sqlbase/system.go#L139)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

Can you document some of the fields as well as discuss why you need the indexes
</blockquote></details>

https://github.com/cockroachdb/cockroach/blob/master/docs/RFCS/web_session_login.md#backend-changes

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17171)*
<!-- Sent from Reviewable.io -->

petermattis commented:
I agree that `sessions` is a bit too general. `web_sessions` or `auth_sessions` sound good to me. 

---

Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17171#-:-Kpp-xaryMnReuRPs3ps:b2xbjz4)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[pkg/sql/sqlbase/system.go, line 139 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17171#-Kpd8KqkP6xcBEgs4p2k-r1-139:-Kppwtqzai_E99dmfNjC:b-y85k9k) ([raw file](https://github.com/cockroachdb/cockroach/blob/81c3b2e4d025f7a44b4d305d54b4ef348ddd8dd2/pkg/sql/sqlbase/system.go#L139)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

https://github.com/cockroachdb/cockroach/blob/master/docs/RFCS/web_session_login.md#backend-changes
</blockquote></details>

I don't have a strong preference either, but I guess I would lean towards consistency with the existing tables. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17171)*
<!-- Sent from Reviewable.io -->

mrtracy commented:
Review status: 4 of 12 files reviewed at latest revision, 3 unresolved discussions.

---

*[pkg/sql/sqlbase/system.go, line 139 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17171#-Kpd8KqkP6xcBEgs4p2k-r1-139:-Kpuy21vt26dca-fqtZF:b-vvet3a) ([raw file](https://github.com/cockroachdb/cockroach/blob/81c3b2e4d025f7a44b4d305d54b4ef348ddd8dd2/pkg/sql/sqlbase/system.go#L139)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

I don't have a strong preference either, but I guess I would lean towards consistency with the existing tables. 
</blockquote></details>

Added a comment with some explanation and directions to the RFC, keeping the camelCase names for consistency, renamed table to "web_sessions".

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17171)*
<!-- Sent from Reviewable.io -->

mrtracy commented:
All comments addressed, PTAL

---

Review status: 1 of 12 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17171#-:-Kq-JCR2Ud02yOUIPVrz:btygt1q)*
<!-- Sent from Reviewable.io -->

mrtracy commented:
@petermattis @bdarnell @benesch PTAL, going to be blocked on this soon

benesch commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>, but I've only reviewed the areas I originally commented on.

---

Reviewed 11 of 11 files at r3.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[pkg/migrations/migrations.go, line 358 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/17171#-KqOzh_MdCkiMhWWXLiN:-KqOzh_MdCkiMhWWXLiO:b-w48v4l) ([raw file](https://github.com/cockroachdb/cockroach/blob/a8bd5021f43597139f5454f40ea8e53c9d1dbaa6/pkg/migrations/migrations.go#L358)):*
> ```Go
> }
> 
> func createSessionsTable(ctx context.Context, r runner) error {
> ```

tiny nit: `createWebSessionsTable`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17171#-:-KqOz_ZbTOf7E17SK3Mo:bd4q6j3)*
<!-- Sent from Reviewable.io -->

petermattis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17171#-:-KqP4p5oq9HP5VXoNPPV:bnfp4nl)*
<!-- Sent from Reviewable.io -->

mrtracy commented:
Review status: 3 of 12 files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---

*[pkg/migrations/migrations.go, line 358 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/17171#-KqOzh_MdCkiMhWWXLiN:-KqR7MKm9vsZaJPiVYbW:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a8bd5021f43597139f5454f40ea8e53c9d1dbaa6/pkg/migrations/migrations.go#L358)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

tiny nit: `createWebSessionsTable`
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17171)*
<!-- Sent from Reviewable.io -->
